src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt

```diff
@@ -764,99 +735,136 @@ private class MacroResolver private constructor() {
-        val parentPsi = (element.psi as? RsExpandedElement)?.expandedFrom ?: parentStub.psi
+        val expandedFrom = (element.psi as? RsExpandedElement)?.expandedFrom
+        if (expandedFrom != null && processExpandedFrom(expandedFrom)) return true
+        val parentPsi = expandedFrom ?: parentStub.psi
         return when {
-            parentPsi is RsFile -> processScopesInLexicalOrderUpward(parentPsi.declaration ?: return false, processor)
+            parentPsi is RsFile -> processScopesInLexicalOrderUpward(parentPsi.declaration ?: return false)
             // Optimization. Let this function be tailrec if go up by stub parent
-            parentPsi != parentStub.psi -> processScopesInLexicalOrderUpward(parentPsi, processor)
-            else -> stubBasedProcessScopesInLexicalOrderUpward(parentStub, processor)
+            parentPsi != parentStub.psi -> processScopesInLexicalOrderUpward(parentPsi)
+            else -> stubBasedProcessScopesInLexicalOrderUpward(parentStub)
         }
     }
 
-    private fun processMacrosAtScopeEntry(
-        item: RsElement,
-        processor: (RsNamedElement) -> Boolean
-    ): Boolean {
-        when (item) {
-            is RsMacro -> if (processor(item)) return true
-            is RsModItem ->
-                if (missingMacroUse.hitOnFalse(item.hasMacroUse)) {
-                    if (processAll(visibleMacros(item), processor)) return true
-                }
-            is RsModDeclItem ->
-                if (missingMacroUse.hitOnFalse(item.hasMacroUse)) {
-                    val mod = item.reference.resolve() as? RsMod ?: return false
-                    if (processAll(visibleMacros(mod), processor)) return true
-                }
-            is RsExternCrateItem -> {
-                val mod = item.reference.resolve() as? RsFile ?: return false
-                if (missingMacroUse.hitOnFalse(item.hasMacroUse)) {
-                    // If extern crate has `#[macro_use]` attribute
-                    // we can use all exported macros from the corresponding crate
-                    if (processAll(exportedMacros(mod), processor)) return true
-                } else {
-                    // otherwise we can use only reexported macros
-                    val reexportedMacros = reexportedMacros(item)
-                    if (reexportedMacros != null) {
-                        // via #[macro_reexport] attribute (old way)
-                        if (processAll(reexportedMacros, processor)) return true
-                    } else {
-                        // or from `use` items (new way)
-                        exportingMacrosCrates[item.nameWithAlias] = mod
-                    }
-                }
-
-
+    // Optimization: macros often expand to calls to themselves. We know that if we are expanded from
+    // a macro call with name "foo", the  macro "foo" is visible at that place. Also such macro
+    // should be already cached
```

Comment:
What do you think about the following version?

Optimization: macros often invoke themselves recursively. If the expansion of  `foo` macro contains `foo` macro call, it means:
* it's the same macro as a calling one
* name resolution result of calling macro should be already cached

So we can just get this result from the cache directly and avoid expensive name resolution in such cases